src/inmanta/server/services/orchestrationservice.py

```diff
@@ -59,6 +62,161 @@
+        self.resource_set = resource_set
+
+    def is_shared_resource(self) -> bool:
+        return self.resource_set is None
+
+
+class PairedResource:
+    def __init__(
+        self,
+        new_resource: ResourceWithResourceSet,
+        old_resource: Optional[ResourceWithResourceSet],
+    ) -> None:
+        self.new_resource = new_resource
+        self.old_resource = old_resource
+
+    def is_update(self) -> bool:
+        if self.old_resource is None:
+            return False
+        attr_names_new_resource = set(self.new_resource.resource).difference("id")
+        attr_names_old_resource = set(self.old_resource.resource).difference("id")
+        return attr_names_new_resource != attr_names_old_resource or any(
+            self.new_resource.resource[k] != self.old_resource.resource[k] for k in attr_names_new_resource
+        )
+
+    def is_new_resource(self) -> bool:
+        return self.old_resource is None
+
+    def resource_changed_resource_set(self) -> bool:
+        assert self.old_resource is not None
+        return self.new_resource.resource_set != self.old_resource.resource_set
+
+
+class PartialUpdateMerger:
+    def __init__(
+        self,
+        partial_updates: List[Dict[str, Any]],
+        resource_sets: Dict[ResourceIdStr, Optional[str]],
+        removed_resource_sets: List[str],
+        env: data.Environment,
+    ) -> None:
+        self.partial_updates = partial_updates
+        self.resource_sets = resource_sets
+        self.removed_resource_sets = removed_resource_sets
+        self.env = env
+        self.paired_resources: Optional[List[PairedResource]] = None
+
+    def pair_resources_partial_update_to_old_version(self, old_resources: Dict[ResourceIdStr, ResourceWithResourceSet]) -> None:
+        """
+        returns a list of paired resources
+
+        :param old_resources: a list of tuples containing the resource in the previous version of the model
+        and its resource_set.
+        :param partial_updates: The list of resources part of the partial compile.
+        """
+        partial_updates: List[Dict[str, Any]] = self.partial_updates
+        paired_resources: List[PairedResource] = []
+        for partial_update in partial_updates:
+            key = Id.parse_id(partial_update["id"]).resource_str()
+            resource_set = self.resource_sets.get(key)
+            pair: PairedResource = PairedResource(ResourceWithResourceSet(partial_update, resource_set), None)
+            if key in old_resources:
+                pair.old_resource = ResourceWithResourceSet(old_resources[key].resource, old_resources[key].resource_set)
+            paired_resources.append(pair)
+        self.paired_resources = paired_resources
+
+    async def _get_old_resources(
+        self,
+    ) -> Tuple[Dict[ResourceIdStr, ResourceWithResourceSet], Dict[ResourceIdStr, Optional[str]]]:
+        old_data = await data.Resource.get_resources_in_latest_version(environment=self.env.id)
```

Comment:
Should I do something about this now or create a ticket?

Reply:
I would detect this scenario and make the `put_partial()` endpoint return a status code 409 (Conflict) if it occurs. The user/compiler-service can then respond to it by running a full compile for example. @sanderr @wouterdb Wdyt?

Reply:
When we discussed this yesterday I agreed with that, but the more I think about it the more I'm leaning towards the lock. I think it should be relatively simple to implement and it's quite a bit more flexible. It allows for out-of-order `put_version` calls because it guards against the write-since-last-read race.
To be clear, I'm not saying I have an objection against the 409, just that I like it a bit less.

Reply:
@FloLey we definitely need to address this as part of this epic, but not necessarily as part of this ticket I think. Creating a new one offers a bit more parallelism so I would go for that. Could you create one once the discussion here has reached some sort of concensus (even if that consensus is that we need to think about it some more)?

Reply:
The problem with the lock-based approach is that it becomes quite complex once you take failure scenarios into account. Given the fact this edge case should be rather rare in practice (The server executes compiles in a strictly linear order), I think the locking approach would make things unnecessarily complex.

Reply:
It's more complex in the sense that:
* We need to introduce an additional API endpoint
* We need to take into account that a client fails while it has the lock, so we need to make this lock timeout as well.
* It's also a breaking change that will make the previous exporter non-compatible with the new server.
As it's currently not clear whether we actually need this feature (i.e. whether this issue causes a lot of trouble in practice), I think it would be good to go for a simple approach and learn from the experiences we obtain using the initial implementation.

I am going to organize a short meeting to discuss this. Otherwise this discussion is going to take forever.

Reply:
@FloLey I would leave this issue as it is for now and move on. I doesn't make sense to block the PR on this.

Reply:
I've just had a talk with Arnaud about this. I could leave our (semi-)conclusions here but I might as well just leave them in a new ticket and link there. So I'll create the tickets.

Reply:
thanks :)

Reply:
I've created #4405 and #4406. The latter is meant as a follow-up ticket to consider at some point in the future (not part of this epic), unless we decide to pick it up now after all. @arnaudsjs with the suggestion I included in the ticket, about dropping reservations for partial compiles only, I actually think it might be feasible (and semantically more clear) to implement it even in the short term. We'll need to discuss this further, then pick up one or the other.